### ISSUE TITLE
feat(gatsby-source-strapi): Create Gatsby data layer types for empty/missing Strapi data

### ISSUE THREAD

**Issue by stephanmax** [gatsby-uc/plugins#489]
## Is your feature request related to a problem? Please describe.

I am running into more and more problems around this [limitation of the plugin](https://github.com/gatsby-uc/plugins/tree/main/packages/gatsby-source-strapi#restrictions-and-limitations):

> At the moment, fields that do not have at least one populated instance will not be created in the GraphQL schema. This can break your site when field values get removed. You may workaround with an extra content entry with all fields filled out.

We have set up a complex Strapi instance that is supposed to power five websites and potentially more in the future. Each website uses different content types that are modeled in Strapi (blog posts, job ads, reports, etc.), but all websites should be able to support all content types.

## Describe the solution you'd like

I am not sure if the limitation above is rooted in Gatsby’s data layer or if it particularly hard to implement, but in my mind the `gatsby-source-strapi` plugin should not be opinionated in that way. Basically, every property of every content type is required. Setting it to `required` or not in Strapi does not have any effect, and missing/empty data is throwing endless errors in Gatsby’s page queries that might not even be used during the building of a particular website.

I would like the plugin to create all types for modeled content types in Strapi, regardless of the presence/absence of actual values. The Gatsby build step with GraphQL queries would then succeed and I can handle missing.empty data in my React components.

## Describe alternatives you've considered

After googling around, two possible workarounds emerged:
- Add dummy date to Strapi (❌ Feels **very** hacky and confuses editors in Strapi.)
- Add manual type information on Gatsby’s side (❌ A lot of cumbersome, manual code needed in `gatsby-node.js`. Also, this approach does not seem to work when you want to account for content types that might not be present at all *or* just have some missing properties.)

Both alternative solutions are not feasible for us.

## Additional context

Could anyone involved in the plugin development so far shed some light on this and most importantly tell me, whether this limitation has any hard reasons to be there? If not and this is just a matter of limited resources to maintain this plugin (and because Gatsby seems to die slowly but surely), let me know as well, and I will try to implement a solution and PR this here. Many thanks!

**Comment by moonmeister** [comment]
I can't speak to the strapi side of this issue but this is not uncommon for source plugins because of how Gatsby works. Because Gatsby can take any source of data and create a graphql API in has to create those graphql types. It does so by default based on the data it receives. this, if it doesn't return a field, it doesn't know it exists. 

If there is a way to fetch the expected schema from Strapi then this plugin could be extended to use that data when typing graphql and thus solving this problem. 

You're welcome to submit a PR, @laurenskling is the only other one who occasionally works on this plugin as well so might be able to give more Strapi specific feedback

**Comment by stephanmax** [comment]
Thanks for your reply, Alex.

If I understand correctly, that means there are two ways how to go about this:

1. Either make Strapi return the full expected schema or
2. Make the plugin play nice with any manual type information on Strapi content types that have been provided on `gatsby-node.js`.

I tried (2) but did not manage to create the correct types without the plugin overwriting crucial type information. Will look into (1) then. Thanks for pinging Laurens, any insight from the Strapi side would be great.

**Comment by laurenskling** [comment]
@stephanmax , yes. These are the two options. Always returning all fields you use in your queries (which is not a realistic method), or manually add the types via `createSchemaCustomization`.

I myself use the second method. When I create new stuff, I load it, see what schema Gatsby creates for it, and copy the exact names into `createTypes` in `createSchemaCustomization`. When you use the same type names, it will mix your manual and loaded fields. Be aware you need to set relations as: `image: STRAPI__MEDIA @link(by: "id", from: "image___NODE")`. 
Yes it's a lot of manual labour. But you only need to set it up once and this is the most stable way of getting it done.

This source plugin was created by Strapi a couple of yours ago. It loads all data from the REST API, thus, not knowing anything about schema's. Typescript might also help, but TS wasn't used yet back then. The choice to load everything via REST and not loading any schema's to create the types, created this situation. 

I guess we might be able to work with `/api/content-type-builder/content-types`, which is used right now to do things like see if it's a singleType or not and determining if a field is a component/DZ/etc. I'm pretty sure we could fully create types from this as well. If you want, you can give it a go. But I'm not sure if such a big change to this plugin is worth while. I don't think there is much future in Gatsby. RIP.

**Comment by stephanmax** [comment]
Thanks for your thoughts and insights on this, Laurens—much appreciated. I will give the manual typing in `createSchemaCustomization` another go; I remember having some struggles understanding the differences of the `StrapiContentType` and `STRAPI_CONTENT_TYPE` types, but maybe I was too hasty. I will try your approach and copy+paste the generated types from the GraphQL playground when running Gatsby build with full dummy data.

**Comment by laurenskling** [comment]
That would be my best suggestion. Good luck.

**Comment by laurenskling** [comment]
related to: https://github.com/gatsby-uc/plugins/issues/433
